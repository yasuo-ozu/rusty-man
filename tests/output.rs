// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: MIT

use std::env;
use std::process;
use std::sync;

use assert_cmd::prelude::*;

static mut DIR: Option<tempfile::TempDir> = None;
static INIT: sync::Once = sync::Once::new();

fn run_cargo_doc() {
    INIT.call_once(|| {
        let dir = tempfile::tempdir().unwrap();
        process::Command::new(env::var_os("CARGO").unwrap())
            .arg("doc")
            .args(&["--package", "anyhow"])
            .args(&["--package", "log"])
            .args(&["--package", "rand_core"])
            .arg("--target-dir")
            .arg(dir.path())
            .output()
            .unwrap();
        unsafe {
            DIR = Some(dir);
        }
    });
}

fn run(args: &[&str]) -> assert_cmd::assert::Assert {
    run_cargo_doc();

    let dir = unsafe { DIR.as_ref().unwrap() };

    process::Command::cargo_bin(env!("CARGO_PKG_NAME"))
        .unwrap()
        .args(&["--no-default-sources", "--source"])
        .arg(dir.path().join("doc"))
        .args(&["--viewer", "plain"])
        .args(&["--width", "100"])
        .args(args)
        .assert()
}

fn get_stdout(args: &[&str]) -> String {
    let cmd = run(args).success().stderr("");
    String::from_utf8(cmd.get_output().stdout.clone()).unwrap()
}

fn assert_doc(item: &str) {
    insta::assert_snapshot!(get_stdout(&[item]));
}

fn assert_examples(item: &str) {
    insta::assert_snapshot!(get_stdout(&["-e", item]));
}

macro_rules! assert_doc {
    ($( $( #[$attrs:meta] )* $name:ident: $item:literal ),* $(,)? ) => {
        $(
            #[test]
            $( #[$attrs] )*
            fn $name() {
                assert_doc($item);
            }
        )*
    };
}

macro_rules! assert_examples {
    ($( $( #[$attrs:meta] )* $name:ident: $item:literal ),* $(,)? ) => {
        $(
            #[test]
            $( #[$attrs] )*
            fn $name() {
                assert_examples($item);
            }
        )*
    };
}

assert_doc![
    test_mod_anyhow: "anyhow",
    test_macro_anyhow_anyhow: "anyhow::anyhow",
    test_macro_anyhow_ensure: "anyhow::ensure",
    #[ignore]
    test_struct_anyhow_error: "anyhow::Error",
    test_trait_anyhow_context: "anyhow::Context",
    test_type_anyhow_result: "anyhow::Result",
];

assert_doc![
    #[ignore]
    test_mod_log: "log",
    #[ignore]
    test_macro_log_debug: "log::debug",
    #[ignore]
    test_struct_log_metadata: "log::Metadata",
    test_enum_log_level: "log::Level",
    #[ignore]
    test_constant_log_static_max_level: "log::STATIC_MAX_LEVEL",
    test_trait_log_log: "log::Log",
    test_fn_log_logger: "log::logger",
    test_fn_log_set_logger_racy: "log::set_logger_racy",
];

assert_doc![
    test_mod_rand_core: "rand_core",
    test_trait_rand_core_rngcore: "rand_core::RngCore",
    test_trait_rand_core_seedablerng: "rand_core::SeedableRng",
    test_struct_rand_core_block_blockrng: "rand_core::block::BlockRng",
];

assert_examples![
    test_examples_mod_anyhow: "anyhow",
    #[ignore]
    test_examples_mod_log: "log",
    test_examples_struct_rand_core_rngcore: "rand_core::RngCore",
];
