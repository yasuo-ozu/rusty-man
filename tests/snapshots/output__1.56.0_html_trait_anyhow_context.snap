---
source: tests/output.rs
expression: "get_stdout(path, &[\"anyhow::Context\"])"

---
anyhow                                  Trait anyhow::Context                              rusty-man

SYNOPSIS
      pub trait Context<T, E>: Sealed {
          fn context<C>(self, context: C) -> Result<T, Error>
          where
              C: Display + Send + Sync + 'static;
          fn with_context<C, F>(self, f: F) -> Result<T, Error>
          where
              C: Display + Send + Sync + 'static,
              F: FnOnce() -> C;
      }

DESCRIPTION
      Provides the `context` method for `Result`.
      
      This trait is sealed and cannot be implemented for types outside of `anyhow`.
      
      
      # Example
      
      `use anyhow::{Context, Result};
      use std::fs;
      use std::path::PathBuf;
      pub struct ImportantThing {
          path: PathBuf,
      }
      impl ImportantThing {
          pub fn detach(&mut self) -> Result<()> {...}
      }
      pub fn do_it(mut it: ImportantThing) -> Result<Vec<u8>> {
          it.detach().context("Failed to detach the important thing")?;
          let path = &it.path;
          let content = fs::read(path)
              .with_context(|| format!("Failed to read instrs from {}", path.display()))?;
          Ok(content)
      }`
      
      When printed, the outermost context would be printed first and the lower level underlying
      causes would be enumerated below.
      
      `Error: Failed to read instrs from ./path/to/instrs.json
      Caused by:
          No such file or directory (os error 2)`
      
      
      
      # Effect on downcasting
      
      After attaching context of type `C` onto an error of type `E`, the resulting `anyhow::Error`
      may be downcast to `C` **or** to `E`.
      
      That is, in codebases that rely on downcasting, Anyhow’s context supports both of the
      following use cases:
      
      * **Attaching context whose type is insignificant onto errors whose type is used in
        downcasts.**
        
        In other error libraries whose context is not designed this way, it can be risky to
        introduce context to existing code because new context might break existing working
        downcasts. In Anyhow, any downcast that worked before adding context will continue to work
        after you add a context, so you should freely add human-readable context to errors wherever
        it would be helpful.
        
        `use anyhow::{Context, Result};
        fn do_it() -> Result<()> {
            helper().context("Failed to complete the work")?;
            ...
        }
        fn main() {
            let err = do_it().unwrap_err();
            if let Some(e) = err.downcast_ref::<SuspiciousError>() {
                // If helper() returned SuspiciousError, this downcast will
                // correctly succeed even with the context in between.
            }
        }`
      * **Attaching context whose type is used in downcasts onto errors whose type is
        insignificant.**
        
        Some codebases prefer to use machine-readable context to categorize lower level errors in a
        way that will be actionable to higher levels of the application.
        
        `use anyhow::{Context, Result};
        fn do_it() -> Result<()> {
            helper().context(HelperFailed)?;
            ...
        }
        fn main() {
            let err = do_it().unwrap_err();
            if let Some(e) = err.downcast_ref::<HelperFailed>() {
                // If helper failed, this downcast will succeed because
                // HelperFailed is the context that has been attached to
                // that error.
            }
        }`

METHODS
   Required Methods
      context
            fn context<C>(self, context: C) -> Result<T, Error> 
            where
                C: Display + Send + Sync + 'static,

            Wrap the error value with additional context.

      with_context
            fn with_context<C, F>(self, f: F) -> Result<T, Error> 
            where
                C: Display + Send + Sync + 'static,
                F: FnOnce() -> C,

            Wrap the error value with additional context that is evaluated lazily only once an error
            does occur.


